File: www/DiceRoller.php
<?php

class DiceRoller
{
    /**
     * Roll a single die with optional exploding
     * 
     * @param int $sides Number of sides on the die
     * @param bool $explode Whether the die should explode on max value
     * @return int The roll result
     */
    private static function rollSingleDie(int $sides, bool $explode = false): int
    {
        $roll = random_int(1, $sides);
        
        // If exploding and rolled max, add another roll
        if ($explode && $roll === $sides) {
            $roll += self::rollSingleDie($sides, true);
        }
        
        return $roll;
    }
    
    /**
     * Roll dice with the specified notation (e.g., "2d6", "1d20+5", "3d8-2", "4d6!")
     * 
     * @param string $notation Dice notation (e.g., "2d6", "1d20+5", "4d6!" for exploding)
     * @return array Result with individual rolls, total, and details
     */
    public static function roll(string $notation): array
    {
        // Parse dice notation: XdY+Z or XdY-Z or XdY or XdY! (exploding)
        $pattern = '/^(\d+)d(\d+)(!)?([+-]\d+)?$/i';
        
        if (!preg_match($pattern, $notation, $matches)) {
            throw new InvalidArgumentException("Invalid dice notation: {$notation}. Use format like '2d6', '1d20+5', or '4d6!' for exploding");
        }
        
        $count = (int)$matches[1];
        $sides = (int)$matches[2];
        $explode = isset($matches[3]) && $matches[3] === '!';
        $modifier = isset($matches[4]) ? (int)$matches[4] : 0;
        
        // Validate inputs
        if ($count < 1 || $count > 100) {
            throw new InvalidArgumentException("Number of dice must be between 1 and 100");
        }
        
        if ($sides < 2 || $sides > 1000) {
            throw new InvalidArgumentException("Number of sides must be between 2 and 1000");
        }
        
        // Roll the dice
        $rolls = [];
        $sum = 0;
        
        for ($i = 0; $i < $count; $i++) {
            $roll = self::rollSingleDie($sides, $explode);
            $rolls[] = $roll;
            $sum += $roll;
        }
        
        // Sort rolls in descending order for easier reading
        rsort($rolls);
        
        $total = $sum + $modifier;
        
        return [
            'notation' => $notation,
            'rolls' => $rolls,
            'sum' => $sum,
            'modifier' => $modifier,
            'total' => $total,
            'exploding' => $explode,
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Roll multiple different dice notations
     * 
     * @param array $notations Array of dice notations
     * @return array Array of results
     */
    public static function rollMultiple(array $notations): array
    {
        $results = [];
        
        foreach ($notations as $notation) {
            try {
                $results[] = self::roll($notation);
            } catch (Exception $e) {
                $results[] = [
                    'notation' => $notation,
                    'error' => $e->getMessage()
                ];
            }
        }
        
        return $results;
    }
    
    /**
     * Roll with advantage (roll twice, take higher)
     * 
     * @param string $notation Dice notation
     * @return array Result with both rolls and the higher value
     */
    public static function rollWithAdvantage(string $notation): array
    {
        $roll1 = self::roll($notation);
        $roll2 = self::roll($notation);
        
        return [
            'notation' => $notation,
            'roll1' => $roll1,
            'roll2' => $roll2,
            'result' => $roll1['total'] >= $roll2['total'] ? $roll1 : $roll2,
            'advantage' => true,
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Roll with disadvantage (roll twice, take lower)
     * 
     * @param string $notation Dice notation
     * @return array Result with both rolls and the lower value
     */
    public static function rollWithDisadvantage(string $notation): array
    {
        $roll1 = self::roll($notation);
        $roll2 = self::roll($notation);
        
        return [
            'notation' => $notation,
            'roll1' => $roll1,
            'roll2' => $roll2,
            'result' => $roll1['total'] <= $roll2['total'] ? $roll1 : $roll2,
            'disadvantage' => true,
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Roll dice with a target number (Shadowrun-style success counting)
     * 
     * @param string $notation Dice notation (e.g., "6d6" or "6d6!")
     * @param int $targetNumber The target number for successes
     * @return array Result with successes counted
     */
    public static function rollWithTargetNumber(string $notation, int $targetNumber): array
    {
        $result = self::roll($notation);
        
        // Count successes and track failures
        $successes = 0;
        $failures = [];
        $allOnes = true;
        
        foreach ($result['rolls'] as $roll) {
            if ($roll >= $targetNumber) {
                $successes++;
            } else {
                $failures[] = $roll;
            }
            
            if ($roll !== 1) {
                $allOnes = false;
            }
        }
        
        return [
            'notation' => $notation,
            'target_number' => $targetNumber,
            'rolls' => $result['rolls'],
            'successes' => $successes,
            'failures' => $failures,
            'failure_count' => count($failures),
            'all_ones' => $allOnes,
            'exploding' => $result['exploding'],
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Opposed roll (two sets of dice rolled against each other)
     * 
     * @param string $notation1 First roller's dice notation
     * @param int $tn1 First roller's target number
     * @param string $notation2 Opponent's dice notation
     * @param int $tn2 Opponent's target number
     * @return array Result with net successes
     */
    public static function rollOpposed(string $notation1, int $tn1, string $notation2, int $tn2): array
    {
        $roll1 = self::rollWithTargetNumber($notation1, $tn1);
        $roll2 = self::rollWithTargetNumber($notation2, $tn2);
        
        $netSuccesses = $roll1['successes'] - $roll2['successes'];
        
        return [
            'roller' => [
                'notation' => $notation1,
                'target_number' => $tn1,
                'rolls' => $roll1['rolls'],
                'successes' => $roll1['successes']
            ],
            'opponent' => [
                'notation' => $notation2,
                'target_number' => $tn2,
                'rolls' => $roll2['rolls'],
                'successes' => $roll2['successes']
            ],
            'net_successes' => $netSuccesses,
            'winner' => $netSuccesses > 0 ? 'roller' : ($netSuccesses < 0 ? 'opponent' : 'tie'),
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Roll initiative (Shadowrun-style - never exploding)
     * 
     * @param string $notation Dice notation (e.g., "2d6+10")
     * @return array Result with initiative score and phases
     */
    public static function rollInitiative(string $notation): array
    {
        // Parse dice notation, removing any ! if present (initiative never explodes)
        $notation = str_replace('!', '', $notation);
        $pattern = '/^(\d+)d(\d+)([+-]\d+)?$/i';
        
        if (!preg_match($pattern, $notation, $matches)) {
            throw new InvalidArgumentException("Invalid initiative notation: {$notation}. Use format like '2d6+10'");
        }
        
        $count = (int)$matches[1];
        $sides = (int)$matches[2];
        $modifier = isset($matches[3]) ? (int)$matches[3] : 0;
        
        // Validate inputs
        if ($count < 1 || $count > 100) {
            throw new InvalidArgumentException("Number of dice must be between 1 and 100");
        }
        
        if ($sides < 2 || $sides > 1000) {
            throw new InvalidArgumentException("Number of sides must be between 2 and 1000");
        }
        
        // Roll the dice (never exploding for initiative)
        $rolls = [];
        $sum = 0;
        
        for ($i = 0; $i < $count; $i++) {
            $roll = random_int(1, $sides);
            $rolls[] = $roll;
            $sum += $roll;
        }
        
        // Sort rolls in descending order
        rsort($rolls);
        
        $initiative = $sum + $modifier;
        
        // Calculate phases (every 10 points)
        $phases = [];
        $currentPhase = $initiative;
        while ($currentPhase > 0) {
            $phases[] = $currentPhase;
            $currentPhase -= 10;
        }
        
        return [
            'notation' => $notation,
            'rolls' => $rolls,
            'modifier' => $modifier,
            'initiative' => $initiative,
            'phases' => $phases,
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Roll with dice pools (Shadowrun-style)
     * Allows tracking multiple pools separately (skill, combat pool, karma pool, etc.)
     * 
     * @param array $pools Array of pools, each with 'name' and 'notation' keys
     * @param int $targetNumber The target number for all pools
     * @return array Results with individual pool breakdowns and totals
     */
    public static function rollWithPools(array $pools, int $targetNumber): array
    {
        if (empty($pools)) {
            throw new InvalidArgumentException("At least one pool must be provided");
        }
        
        $poolResults = [];
        $totalSuccesses = 0;
        $totalFailures = 0;
        $allFailures = [];
        $summaryParts = [];
        
        foreach ($pools as $pool) {
            if (!isset($pool['name']) || !isset($pool['notation'])) {
                throw new InvalidArgumentException("Each pool must have 'name' and 'notation' keys");
            }
            
            // Roll this pool
            $result = self::rollWithTargetNumber($pool['notation'], $targetNumber);
            
            $poolResult = [
                'name' => $pool['name'],
                'notation' => $pool['notation'],
                'rolls' => $result['rolls'],
                'successes' => $result['successes'],
                'failures' => $result['failures'],
                'failure_count' => $result['failure_count'],
                'exploding' => $result['exploding']
            ];
            
            $poolResults[] = $poolResult;
            $totalSuccesses += $result['successes'];
            $totalFailures += $result['failure_count'];
            $allFailures = array_merge($allFailures, $result['failures']);
            $summaryParts[] = "{$pool['name']}: {$result['successes']}";
        }
        
        $summaryParts[] = "Total: {$totalSuccesses}";
        
        return [
            'target_number' => $targetNumber,
            'pools' => $poolResults,
            'total_successes' => $totalSuccesses,
            'total_failures' => $totalFailures,
            'all_failures' => $allFailures,
            'summary' => implode(', ', $summaryParts),
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Opposed roll with dice pools (Shadowrun-style)
     * General-purpose opposed test - no hardcoded logic for specific test types
     * 
     * @param array $side1Pools Array of pools for side 1
     * @param int $side1TN Target number for side 1
     * @param array $side2Pools Array of pools for side 2
     * @param int $side2TN Target number for side 2
     * @param string $side1Label Label for side 1 (e.g., "Attacker", "Sneaker")
     * @param string $side2Label Label for side 2 (e.g., "Defender", "Watcher")
     * @return array Complete results for both sides with net successes
     */
    public static function rollOpposedWithPools(
        array $side1Pools,
        int $side1TN,
        array $side2Pools,
        int $side2TN,
        string $side1Label = 'Side 1',
        string $side2Label = 'Side 2'
    ): array
    {
        $side1Result = self::rollWithPools($side1Pools, $side1TN);
        $side2Result = self::rollWithPools($side2Pools, $side2TN);
        
        $netSuccesses = $side1Result['total_successes'] - $side2Result['total_successes'];
        
        $winner = 'tie';
        if ($netSuccesses > 0) {
            $winner = $side1Label;
        } elseif ($netSuccesses < 0) {
            $winner = $side2Label;
        }
        
        return [
            'test_type' => 'opposed_pools',
            'side1' => [
                'label' => $side1Label,
                'target_number' => $side1TN,
                'pools' => $side1Result['pools'],
                'total_successes' => $side1Result['total_successes'],
                'summary' => $side1Result['summary']
            ],
            'side2' => [
                'label' => $side2Label,
                'target_number' => $side2TN,
                'pools' => $side2Result['pools'],
                'total_successes' => $side2Result['total_successes'],
                'summary' => $side2Result['summary']
            ],
            'net_successes' => $netSuccesses,
            'winner' => $winner,
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Re-roll failed dice using Karma Pool (Shadowrun-style)
     * Cost escalates: 1st re-roll = 1 Karma, 2nd = 2 Karma, 3rd = 3 Karma, etc.
     * 
     * @param array $failedDice Array of failed die values to re-roll
     * @param int $targetNumber The target number for successes
     * @param int $sides Number of sides on the die (default 6)
     * @param bool $exploding Whether dice explode on max value
     * @param int $rerollIteration Which re-roll attempt (1st, 2nd, 3rd, etc.)
     * @return array Re-roll results with new successes and failures
     */
    public static function rerollFailures(
        array $failedDice,
        int $targetNumber,
        int $sides = 6,
        bool $exploding = true,
        int $rerollIteration = 1
    ): array
    {
        if (empty($failedDice)) {
            throw new InvalidArgumentException("No failed dice to re-roll");
        }
        
        if ($rerollIteration < 1) {
            throw new InvalidArgumentException("Re-roll iteration must be at least 1");
        }
        
        $karmaCost = $rerollIteration;
        $rerolledValues = [];
        $newSuccesses = 0;
        $newFailures = [];
        
        // Re-roll each failed die
        foreach ($failedDice as $failedValue) {
            $newRoll = self::rollSingleDie($sides, $exploding);
            $rerolledValues[] = $newRoll;
            
            if ($newRoll >= $targetNumber) {
                $newSuccesses++;
            } else {
                $newFailures[] = $newRoll;
            }
        }
        
        // Sort for readability
        rsort($rerolledValues);
        rsort($newFailures);
        
        return [
            'reroll_iteration' => $rerollIteration,
            'karma_cost' => $karmaCost,
            'next_reroll_cost' => $rerollIteration + 1,
            'failed_dice_count' => count($failedDice),
            'rerolled_values' => $rerolledValues,
            'new_successes' => $newSuccesses,
            'new_failures' => $newFailures,
            'new_failure_count' => count($newFailures),
            'summary' => "Re-rolled " . count($failedDice) . " failures for {$karmaCost} Karma. Got {$newSuccesses} new success" . ($newSuccesses !== 1 ? 'es' : '') . ". " . count($newFailures) . " failure" . (count($newFailures) !== 1 ? 's' : '') . " remain.",
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Avoid disaster when all dice come up 1 (Rule of One)
     * Costs 1 Karma to convert critical failure to simple failure
     * No re-roll is allowed after using this
     * 
     * @param array $rollResult The original roll result with all_ones = true
     * @return array Result of avoiding the disaster
     */
    public static function avoidDisaster(array $rollResult): array
    {
        if (!isset($rollResult['all_ones']) || !$rollResult['all_ones']) {
            throw new InvalidArgumentException("Can only avoid disaster when all dice are 1s");
        }
        
        return [
            'disaster_avoided' => true,
            'karma_cost' => 1,
            'original_result' => 'All 1s - Critical Failure',
            'new_result' => 'Simple Failure',
            'note' => 'Cannot re-roll after avoiding disaster',
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Buy additional dice using Karma Pool
     * 1 Karma per die, up to a maximum (usually skill/attribute level)
     * 
     * @param int $karmaDiceCount Number of Karma dice to buy
     * @param int $targetNumber The target number for successes
     * @param int $sides Number of sides on the die (default 6)
     * @param bool $exploding Whether dice explode on max value
     * @param int|null $maxAllowed Optional cap on Karma dice (usually skill level)
     * @return array Results of the Karma dice roll
     */
    public static function buyKarmaDice(
        int $karmaDiceCount,
        int $targetNumber,
        int $sides = 6,
        bool $exploding = true,
        ?int $maxAllowed = null
    ): array
    {
        if ($karmaDiceCount < 1) {
            throw new InvalidArgumentException("Must buy at least 1 Karma die");
        }
        
        if ($maxAllowed !== null && $karmaDiceCount > $maxAllowed) {
            throw new InvalidArgumentException("Cannot buy more than {$maxAllowed} Karma dice (skill/attribute cap)");
        }
        
        $rolls = [];
        $successes = 0;
        $failures = [];
        
        for ($i = 0; $i < $karmaDiceCount; $i++) {
            $roll = self::rollSingleDie($sides, $exploding);
            $rolls[] = $roll;
            
            if ($roll >= $targetNumber) {
                $successes++;
            } else {
                $failures[] = $roll;
            }
        }
        
        rsort($rolls);
        rsort($failures);
        
        return [
            'karma_dice_bought' => $karmaDiceCount,
            'karma_cost' => $karmaDiceCount,
            'target_number' => $targetNumber,
            'rolls' => $rolls,
            'successes' => $successes,
            'failures' => $failures,
            'failure_count' => count($failures),
            'note' => 'Add these results to your main roll',
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Buy raw successes using Karma Pool
     * Requires at least 1 natural success in the original roll
     * This Karma is PERMANENTLY spent and does not refresh
     * 
     * @param int $currentSuccesses Number of successes already achieved
     * @param int $successesToBuy Number of successes to purchase
     * @return array Result of buying successes
     */
    public static function buySuccesses(
        int $currentSuccesses,
        int $successesToBuy
    ): array
    {
        if ($currentSuccesses < 1) {
            throw new InvalidArgumentException("Must have at least 1 natural success to buy additional successes");
        }
        
        if ($successesToBuy < 1) {
            throw new InvalidArgumentException("Must buy at least 1 success");
        }
        
        $karmaCost = $successesToBuy;
        $newTotal = $currentSuccesses + $successesToBuy;
        
        return [
            'successes_bought' => $successesToBuy,
            'karma_cost' => $karmaCost,
            'permanent_karma_spent' => true,
            'original_successes' => $currentSuccesses,
            'new_total_successes' => $newTotal,
            'warning' => 'This Karma is permanently spent and does not refresh with the next encounter',
            'timestamp' => date('c')
        ];
    }
    
    /**
     * Track initiative for multiple characters (Shadowrun-style)
     * 
     * @param array $characters Array of characters with 'name' and 'notation' keys
     * @return array Complete initiative tracking with phase breakdown
     */
    public static function trackInitiative(array $characters): array
    {
        $results = [];
        
        // Roll initiative for each character
        foreach ($characters as $character) {
            if (!isset($character['name']) || !isset($character['notation'])) {
                throw new InvalidArgumentException("Each character must have 'name' and 'notation' keys");
            }
            
            $roll = self::rollInitiative($character['notation']);
            $results[] = [
                'name' => $character['name'],
                'notation' => $character['notation'],
                'rolls' => $roll['rolls'],
                'modifier' => $roll['modifier'],
                'initiative' => $roll['initiative'],
                'phases' => $roll['phases']
            ];
        }
        
        // Sort by initiative (descending), then by modifier (descending) for tie-breaking
        usort($results, function($a, $b) {
            if ($a['initiative'] === $b['initiative']) {
                return $b['modifier'] - $a['modifier'];
            }
            return $b['initiative'] - $a['initiative'];
        });
        
        // Build phase-by-phase breakdown
        $phaseMap = [];
        foreach ($results as $character) {
            foreach ($character['phases'] as $phase) {
                if (!isset($phaseMap[$phase])) {
                    $phaseMap[$phase] = [];
                }
                $phaseMap[$phase][] = $character;
            }
        }
        
        // Sort phases in descending order
        krsort($phaseMap);
        
        // Build phase order array
        $phaseOrder = [];
        foreach ($phaseMap as $phase => $actors) {
            // Sort actors within phase by initiative, then modifier
            usort($actors, function($a, $b) {
                if ($a['initiative'] === $b['initiative']) {
                    return $b['modifier'] - $a['modifier'];
                }
                return $b['initiative'] - $a['initiative'];
            });
            
            $phaseOrder[] = [
                'phase' => $phase,
                'actors' => array_map(function($actor) {
                    return $actor['name'];
                }, $actors)
            ];
        }
        
        // Build turn summary
        $turnSummary = [];
        foreach ($phaseOrder as $phaseData) {
            $phase = $phaseData['phase'];
            $actorNames = $phaseData['actors'];
            
            // Check for ties in this phase
            $tieNote = '';
            if (count($actorNames) > 1) {
                // Find the actors in this phase from results
                $phaseActors = array_filter($results, function($char) use ($actorNames) {
                    return in_array($char['name'], $actorNames);
                });
                
                // Check if any have the same initiative
                $initiatives = array_map(function($char) {
                    return $char['initiative'];
                }, $phaseActors);
                
                if (count($initiatives) !== count(array_unique($initiatives))) {
                    $tieNote = ' (ties broken by higher modifier)';
                }
            }
            
            $turnSummary[] = "Phase {$phase}: " . implode(', ', $actorNames) . $tieNote;
        }
        
        return [
            'characters' => $results,
            'phase_order' => $phaseOrder,
            'turn_summary' => implode("\n", $turnSummary),
            'timestamp' => date('c')
        ];
    }
}

File: lib/dice_roller.py
#!/usr/bin/env python3
"""
Shadowrun Dice Roller
Implements Shadowrun 2nd Edition dice mechanics including:
- Exploding 6s (Rule of 6)
- Target number success counting
- Opposed rolls
- Initiative tracking
- Karma pool mechanics
"""

import random
from typing import Dict, List, Tuple, Optional, Any
from dataclasses import dataclass


@dataclass
class DiceRoll:
    """Result of a dice roll"""
    rolls: List[int]
    successes: int
    target_number: int
    pool_size: int
    notation: str
    all_ones: bool = False
    critical_glitch: bool = False


class DiceRoller:
    """Shadowrun dice rolling engine"""
    
    @staticmethod
    def parse_notation(notation: str) -> Tuple[int, int, int, bool]:
        """
        Parse dice notation like "6d6+2" or "4d6!"
        Returns: (count, sides, modifier, exploding)
        """
        notation = notation.strip().lower()
        exploding = notation.endswith('!')
        if exploding:
            notation = notation[:-1]
        
        # Split on 'd'
        parts = notation.split('d')
        if len(parts) != 2:
            raise ValueError(f"Invalid dice notation: {notation}")
        
        count = int(parts[0])
        
        # Check for modifier
        modifier = 0
        sides_part = parts[1]
        if '+' in sides_part:
            sides, mod = sides_part.split('+')
            sides = int(sides)
            modifier = int(mod)
        elif '-' in sides_part:
            sides, mod = sides_part.split('-')
            sides = int(sides)
            modifier = -int(mod)
        else:
            sides = int(sides_part)
        
        return count, sides, modifier, exploding
    
    @staticmethod
    def roll_single_die(sides: int = 6) -> int:
        """Roll a single die"""
        return random.randint(1, sides)
    
    @staticmethod
    def roll_with_target_number(
        pool: int,
        target_number: int,
        sides: int = 6,
        exploding: bool = True
    ) -> DiceRoll:
        """
        Roll dice pool against target number (Shadowrun style)
        
        Args:
            pool: Number of dice to roll
            target_number: Target number for success
            sides: Number of sides on each die (default 6)
            exploding: Whether 6s explode (Rule of 6)
        
        Returns:
            DiceRoll object with results
        """
        rolls = []
        successes = 0
        
        for _ in range(pool):
            roll = DiceRoller.roll_single_die(sides)
            combined_value = roll
            
            # Exploding dice (Rule of 6) - combine into single value
            if exploding:
                while roll == sides:
                    roll = DiceRoller.roll_single_die(sides)
                    combined_value += roll
            
            # Add the combined value (e.g., 6+4=10)
            rolls.append(combined_value)
            
            # Check for success on combined value
            if combined_value >= target_number:
                successes += 1
        
        # Check for all ones (critical glitch)
        all_ones = all(r == 1 for r in rolls)
        
        return DiceRoll(
            rolls=rolls,
            successes=successes,
            target_number=target_number,
            pool_size=pool,
            notation=f"{pool}d{sides}{'!' if exploding else ''}",
            all_ones=all_ones,
            critical_glitch=all_ones and successes == 0
        )
    
    @staticmethod
    def roll_opposed(
        pool1: int,
        tn1: int,
        pool2: int,
        tn2: int,
        sides: int = 6,
        exploding: bool = True
    ) -> Dict[str, Any]:
        """
        Roll opposed test between two sides
        
        Returns:
            Dictionary with both rolls and net successes
        """
        roll1 = DiceRoller.roll_with_target_number(pool1, tn1, sides, exploding)
        roll2 = DiceRoller.roll_with_target_number(pool2, tn2, sides, exploding)
        
        net_successes = roll1.successes - roll2.successes
        
        if net_successes > 0:
            winner = "side1"
        elif net_successes < 0:
            winner = "side2"
        else:
            winner = "tie"
        
        return {
            "side1": {
                "pool": pool1,
                "target_number": tn1,
                "rolls": roll1.rolls,
                "successes": roll1.successes,
                "all_ones": roll1.all_ones
            },
            "side2": {
                "pool": pool2,
                "target_number": tn2,
                "rolls": roll2.rolls,
                "successes": roll2.successes,
                "all_ones": roll2.all_ones
            },
            "net_successes": abs(net_successes),
            "winner": winner
        }
    
    @staticmethod
    def roll_initiative(notation: str) -> Dict[str, Any]:
        """
        Roll initiative (never exploding in Shadowrun)
        
        Args:
            notation: Dice notation like "2d6+10"
        
        Returns:
            Dictionary with initiative score and breakdown
        """
        count, sides, modifier, _ = DiceRoller.parse_notation(notation)
        
        rolls = [DiceRoller.roll_single_die(sides) for _ in range(count)]
        total = sum(rolls) + modifier
        
        # Calculate initiative passes (every 10 points = 1 pass)
        passes = 1 + (total // 10)
        
        return {
            "notation": notation,
            "rolls": rolls,
            "modifier": modifier,
            "total": total,
            "passes": passes,
            "breakdown": f"{'+'.join(map(str, rolls))} + {modifier} = {total}"
        }
    
    @staticmethod
    def track_initiative(characters: List[Dict[str, str]]) -> Dict[str, Any]:
        """
        Track initiative for multiple characters
        
        Args:
            characters: List of dicts with 'name' and 'notation' keys
        
        Returns:
            Dictionary with sorted initiative order and phase breakdown
        """
        results = []
        
        for char in characters:
            name = char.get('name', 'Unknown')
            notation = char.get('notation', '1d6')
            
            init_roll = DiceRoller.roll_initiative(notation)
            results.append({
                "name": name,
                "notation": notation,
                "total": init_roll['total'],
                "passes": init_roll['passes'],
                "rolls": init_roll['rolls'],
                "modifier": init_roll['modifier']
            })
        
        # Sort by initiative total (highest first)
        results.sort(key=lambda x: x['total'], reverse=True)
        
        # Build phase breakdown
        max_passes = max(r['passes'] for r in results)
        phases = []
        
        for phase in range(1, max_passes + 1):
            phase_order = [
                r['name'] for r in results 
                if r['passes'] >= phase
            ]
            phases.append({
                "phase": phase,
                "active_characters": phase_order
            })
        
        return {
            "initiative_order": results,
            "phases": phases,
            "max_passes": max_passes
        }
    
    @staticmethod
    def roll_with_pools(
        pools: List[Dict[str, str]],
        target_number: int,
        sides: int = 6,
        exploding: bool = True
    ) -> Dict[str, Any]:
        """
        Roll multiple dice pools (skill, combat pool, karma pool, etc.)
        
        Args:
            pools: List of dicts with 'name' and 'notation' keys
            target_number: Target number for all pools
            sides: Die sides
            exploding: Whether dice explode
        
        Returns:
            Dictionary with individual pool results and totals
        """
        pool_results = []
        total_successes = 0
        all_rolls = []
        
        for pool_def in pools:
            name = pool_def.get('name', 'Unknown Pool')
            notation = pool_def.get('notation', '0d6')
            
            count, _, modifier, _ = DiceRoller.parse_notation(notation)
            
            if count > 0:
                roll = DiceRoller.roll_with_target_number(
                    count, target_number, sides, exploding
                )
                
                pool_results.append({
                    "name": name,
                    "notation": notation,
                    "rolls": roll.rolls,
                    "successes": roll.successes,
                    "all_ones": roll.all_ones
                })
                
                total_successes += roll.successes
                all_rolls.extend(roll.rolls)
        
        return {
            "pools": pool_results,
            "total_successes": total_successes,
            "target_number": target_number,
            "all_rolls": all_rolls
        }
    
    @staticmethod
    def buy_karma_dice(
        karma_dice_count: int,
        target_number: int,
        sides: int = 6,
        exploding: bool = True,
        max_allowed: Optional[int] = None
    ) -> Dict[str, Any]:
        """
        Buy karma dice (costs 1 karma per die, max usually = karma pool rating)
        
        Args:
            karma_dice_count: Number of karma dice to buy
            target_number: Target number for successes
            sides: Die sides
            exploding: Whether dice explode
            max_allowed: Maximum karma dice allowed (optional)
        
        Returns:
            Dictionary with karma dice results
        """
        if max_allowed and karma_dice_count > max_allowed:
            return {
                "error": f"Cannot buy {karma_dice_count} karma dice (max: {max_allowed})",
                "karma_cost": 0,
                "successes": 0
            }
        
        roll = DiceRoller.roll_with_target_number(
            karma_dice_count, target_number, sides, exploding
        )
        
        return {
            "karma_dice_bought": karma_dice_count,
            "karma_cost": karma_dice_count,
            "rolls": roll.rolls,
            "successes": roll.successes,
            "target_number": target_number,
            "all_ones": roll.all_ones
        }
    
    @staticmethod
    def buy_successes(
        current_successes: int,
        successes_to_buy: int
    ) -> Dict[str, Any]:
        """
        Buy successes with karma (2 karma per success)
        
        Args:
            current_successes: Current number of successes
            successes_to_buy: Number of successes to buy
        
        Returns:
            Dictionary with new total and karma cost
        """
        karma_cost = successes_to_buy * 2
        new_total = current_successes + successes_to_buy
        
        return {
            "current_successes": current_successes,
            "successes_bought": successes_to_buy,
            "karma_cost": karma_cost,
            "new_total": new_total
        }
    
    @staticmethod
    def reroll_failures(
        failed_dice: List[int],
        target_number: int,
        sides: int = 6,
        exploding: bool = True,
        iteration: int = 1
    ) -> Dict[str, Any]:
        """
        Reroll failed dice (e.g., with karma or second chance)
        
        Args:
            failed_dice: List of failed die results
            target_number: Target number for success
            sides: Die sides
            exploding: Whether dice explode
            iteration: Which reroll iteration this is
        
        Returns:
            Dictionary with reroll results
        """
        rerolls = []
        new_successes = 0
        still_failed = []
        
        for _ in failed_dice:
            roll = DiceRoller.roll_single_die(sides)
            rerolls.append(roll)
            
            if roll >= target_number:
                new_successes += 1
            else:
                still_failed.append(roll)
            
            # Handle exploding
            if exploding:
                while roll == sides:
                    roll = DiceRoller.roll_single_die(sides)
                    rerolls.append(roll)
                    if roll >= target_number:
                        new_successes += 1
        
        return {
            "reroll_iteration": iteration,
            "dice_rerolled": len(failed_dice),
            "rerolls": rerolls,
            "new_successes": new_successes,
            "still_failed": still_failed,
            "target_number": target_number
        }
    
    @staticmethod
    def avoid_disaster(roll_result: Dict[str, Any]) -> Dict[str, Any]:
        """
        Avoid critical glitch by spending karma (all 1s rolled)
        
        Args:
            roll_result: Previous roll result with all_ones=True
        
        Returns:
            Dictionary indicating if disaster was avoided
        """
        if not roll_result.get('all_ones', False):
            return {
                "error": "No critical glitch to avoid",
                "karma_cost": 0,
                "disaster_avoided": False
            }
        
        return {
            "disaster_avoided": True,
            "karma_cost": 1,
            "message": "Spent 1 karma to avoid critical glitch"
        }
